DTLS 1.3 stateless server ClientHello parsing

julek-wolfssl opened the pull request.

dgarske commented on wolfssl/internal.h:
```diff
@@ -2143,6 +2151,9 @@ WOLFSSL_LOCAL void InitSuites(Suites* suites, ProtocolVersion pv, int keySz,
                               word16 haveFalconSig, word16 haveDilithiumSig,
                               word16 haveAnon, word16 haveNull, int side);
 
+typedef struct TLSX TLSX;
```
```[all-gcc-c99] [4 of 18] [e6f59882fb]
    configure...   real 0m16.867s  user 0m11.762s  sys 0m6.888s
    build...In file included from wolfcrypt/src/asn.c:183:
./wolfssl/internal.h:2621:3: error: redefinition of typedef ‘TLSX’ [-Werror=pedantic]
 2621 | } TLSX;
      |   ^~~~
./wolfssl/internal.h:2154:21: note: previous declaration of ‘TLSX’ was here
 2154 | typedef struct TLSX TLSX;
      |                     ^~~~
```

dgarske commented on wolfssl/internal.h:
```diff
@@ -2944,11 +2964,18 @@ int TLSX_EncryptThenMac_Respond(WOLFSSL* ssl);
+    #ifdef _MSC_VER
+    #pragma warning(disable: 4200)
+    #endif
+    byte   data[];
```
@douzzer , don't we have a better way to do this?

douzzer replied:
@dgarske, @julek-wolfssl already merged in this suppression later in internal.h, in b2a80cbe65.  At a minimum, he should move that later one higher in internal.h so it's done only once.

In wolfSentry I avoid suppressions with this in `wolfsentry_settings.h`:
```
#if defined(__STRICT_ANSI__)
#define WOLFSENTRY_FLEXIBLE_ARRAY_SIZE 1
#elif defined(__GNUC__) && !defined(__clang__)
#define WOLFSENTRY_FLEXIBLE_ARRAY_SIZE
#else
#define WOLFSENTRY_FLEXIBLE_ARRAY_SIZE 0
#endif
```
but the lowest impact way forward is to keep the MSVC suppression and just move it up.

julek-wolfssl replied:
Using the wolfSentry method would use one additional byte wouldn't it? Its not much but it can start to add up.

dgarske reviewed (changes requested):
Also the make check for `./configure --enable-dtls` is failing.

julek-wolfssl committed:
dtls 1.3: Stateless ClientHello parsing

rizlik reviewed (changes requested):
first pass, I'll do a more detailed review later

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+        ret = CheckDtlsCookie(ssl, &ch, isTls13, &cookieGood);
+        if (ret != 0)
+            return ret;
+        if (!cookieGood)
```
The alert is checked and sent in `SendStatelessReply`. I'll refactor to move it closer to the check.

julek-wolfssl replied:
```
https://www.rfc-editor.org/rfc/rfc6347#section-4.2.1
   If a server receives a ClientHello with an invalid cookie, it SHOULD
   treat it the same as a ClientHello with no cookie.  This avoids
   race/deadlock conditions if the client somehow gets a bad cookie
   (e.g., because the server changes its cookie signing key).

   Note to implementors: This may result in clients receiving multiple
   HelloVerifyRequest messages with different cookies.  Clients SHOULD
   handle this by sending a new ClientHello with a cookie in response to
   the new HelloVerifyRequest.
```
No alert for 1.2.

rizlik replied:
no, in 1.2 we should keep sending cookies

rizlik replied:
you didn't move the check closer in the end.

julek-wolfssl replied:
Moving it closer in next push.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -133,45 +134,82 @@ static int ReadVector16(const byte* input, WolfSSL_ConstVector* v)
+            return BUFFER_E;
+        ret = TlsCheckCookie(ssl, ch->cookie.elements + OPAQUE16_LEN,
+                ch->cookie.size - OPAQUE16_LEN);
+        if (ret < 0 && ret != HRR_COOKIE_ERROR)
```
We return all errors except `HRR_COOKIE_ERROR` here.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -183,6 +221,7 @@ static int ParseClientHello(const byte* input, word32 helloSz, WolfSSL_CH* ch)
     if (OPAQUE16_LEN + RAN_LEN + OPAQUE8_LEN > helloSz)
         return BUFFER_ERROR;
 
+    ch->msg = input - DTLS_HANDSHAKE_HEADER_SZ;
```
I don't think so because this is called inside `DoClientHello` and `DoTls13ClientHello` which receive the buffer already advanced past the header. Do you have any other ideas?

rizlik replied:
You can consider removing the `msg` field from the `WolfSSL_CH` struct. The struct represents the parsed client hello (CH) message, and adding a pointer to the original handshake header may not align with the purpose of the struct.
You may also consider changing its name to `raw` or `handshakeHeader`.
If you remove it you can use an additional parameter to pass the handshake information to `SendStatelessReply`. In both case, I think it's better to set/compute the handshake header address in `DoClientHelloStateless` as you can use `inOutIdx` to be sure to not underflow the input buffer.

julek-wolfssl replied:
Changed to use a `raw` pointer and now `Dtls13HashClientHello` recreates the header.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+    if (ret != 0)
+        return ret;
+    if (tlsxSupportedVersions.size == 0) {
+        *isTls13 = 0;
```
What do you mean? Check if the session object has the same version?

rizlik replied:
yes

julek-wolfssl replied:
There is already a check right after
```
    if (pskInfo->isValid && (pskInfo->pv.major != pv.major ||
                             pskInfo->pv.minor != pv.minor))
        return VERSION_ERROR;
```

rizlik replied:
I see, I was referring to the scenario below:
- TlsResumptionIsValid is invoked and returns a valid "session" in pskInfo
- There are no VersionSupported extensions (`tlsxSupportedVersion.size == 0` above).
But upon further consideration, it is not necessary to check the version in DTLS version 1.2 or lower because the server does not have to reply with a supported version extension.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                    doKE = 1;
+                }
+                else {
+                    if ((modes & (1 << PSK_KE)) == 0)
```
Then we prefer the DHE to maintain PFS.

rizlik replied:
In this case, do we still want to error out if `!haveKS` ?

julek-wolfssl replied:
It is an else if the connection can have `PSK_DHE_KE` we choose that.

rizlik replied:
If we have both modes, but the peer didn't provide a keyshare extension (`!haveKS` above)  we error out. In theory, we could have used the `PKS_KE` mode.
Also, is the KeyShare extension obligatory (but with allowed zero-size array)? I can't find a clear answer in the RFC.

julek-wolfssl replied:
For now maybe we should be more cautious. This is also the same logic in our stateful parsing so I was trying to mirror that as much as possible.
```
https://www.rfc-editor.org/rfc/rfc8446#section-4.2.8
   In the ClientHello message, the "extension_data" field of this
   extension contains a "KeyShareClientHello" value:

      struct {
          KeyShareEntry client_shares<0..2^16-1>;
      } KeyShareClientHello;

   client_shares:  A list of offered KeyShareEntry values in descending
      order of client preference.

   This vector MAY be empty if the client is requesting a
   HelloRetryRequest.
```

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+             * found a common group. */
+            if (cs.clientKSE != NULL
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+                    && !usePSK
```
In HRR we can't send keyshare when we don't want to negotiate the curve. Only when we have to negotiate we should send the keyshare.

rizlik replied:
but if using psk with dhe we still want to negotiate the curve

julek-wolfssl replied:
Yes but we do that later in the stateful CH. I don't think we can negotiate that here.

rizlik replied:
In the stateful CH it will be too late. We have to ask the client for a good KS value now as we can't send two HelloRetryRequest in the same connection.

julek-wolfssl replied:
This has been refactored to depend on a `cs.doHelloRetry` variable. Take another look if it makes sense.

julek-wolfssl commented on src/tls.c:
```diff
@@ -6040,46 +6041,40 @@ static int TLSX_SupportedVersions_Parse(WOLFSSL* ssl, const byte* input,
+            /* TLS v1.3 requires supported version extension */
+            if (exts != NULL &&
+                    TLSX_Find(*exts, TLSX_SUPPORTED_VERSIONS) == NULL) {
+                ret = TLSX_Push(exts,
```
To set the `(*exts)->resp = 1;` so that it gets sent in a reply.

rizlik commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+            ret = TLSX_Push(&parsedExts,
+                      TLSX_SUPPORTED_VERSIONS, ssl, ssl->heap);
+            if (ret != 0)
+                goto dtls13_cleanup;
```
If we encounter an error here, do we reset the hashes? 
It would be nice to compute the cookie without changing `ssl->hsHashes`

rizlik replied:
also, Could you consider refactoring the dtls 1.3  part into a separate function? 
Even if refactored out the function seems quite large. It might be helpful to break it up into smaller functions to make the code easier to read.

julek-wolfssl replied:
Refactored to not touch `ssl->hsHashes` while in stateless mode. Refactored 1.3 portion into separate function.

rizlik commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+#endif
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            if (usePSK) {
```
Maybe I'm missing it, but when do we set `usePSK`?

julek-wolfssl replied:
You're right. I forgot to set it. It should be after the `TlsxFindByType(&tlsx, TLSX_PRE_SHARED_KEY, ch->extension)`.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                }
+                else {
+                    /* Only support the default ciphersuite for PSK */
+                    cs.cipherSuite0 = TLS13_BYTE;
```
The PSK does not have to be a ticket. It can also be an identity provided to the user so that they can supply their own key material. Here we assume that the PSK is valid and offer the default PSK cipher suite. This mimics the logic in our stateful logic (obviously there we check if the PSK is valid).

rizlik replied:
So even if `pskInfo->isValid` is false we still have a valid hardcoded/predefined "psk"? This is a little confusing. 
 
In that case, IIRC the ciphersuite is returned by one version of the psk callback. Should we use that value?

julek-wolfssl replied:
Good feedback. I implemented asking the callback and falling back to a full handshake in the last commit.

julek-wolfssl replied:
Good feedback. I implemented asking the callback and falling back to a full handshake in the last commit.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
-    {
-        byte isValid = 0;
-        ret = TlsResumptionIsValid(ssl, &ch, &isValid);
+    ret = TlsResumptionIsValid(ssl, &ch, &pskInfo);
```
I don't think it depends on the TLS 1.3 check. Is there anything specific you have in mind?

rizlik replied:
IIRC now we always do a cookie check in dtls 1.3, we do not have the `DTLS_NO_HVR_ON_RESUME` feature in dtls 1.3. 
If, for other reasons, we always need to check about the session in dtls 1.3 this shouldn't be guarded by `WOLFSSL_DTLS_NO_HVR_ON_RESUME`.
If the logic from 1.3 and 1.2 starts diverging too much it may be worth considering refactoring the function in two to make it simpler.

julek-wolfssl replied:
What do you think of using `WOLFSSL_DTLS_NO_HVR_ON_RESUME` in 1.3 to mean no HRR on resume?

julek-wolfssl replied:
I think extending `WOLFSSL_DTLS_NO_HVR_ON_RESUME` to 1.3 to mean the same thing as 1.2 is easier to understand. This is disabled by default so the user must declare it explicitly anyway.

julek-wolfssl commented on src/tls.c:
```diff
@@ -6028,7 +6029,7 @@ static int TLSX_SupportedVersions_Parse(WOLFSSL* ssl, const byte* input,
 #endif
 
-            if (major != pv.major)
+            if (major != ssl->ctx->method->version.major)
```
This is the logic that was present before. I wanted to maintain as much of the same logic as possible.

julek-wolfssl commented on src/tls.c:
```diff
@@ -6089,7 +6084,7 @@ static int TLSX_SupportedVersions_Parse(WOLFSSL* ssl, const byte* input,
         minor = input[OPAQUE8_LEN];
 
-        if (major != pv.major) {
+        if (major != ssl->ctx->method->version.major) {
```
This is the logic that was present before. I wanted to maintain as much of the same logic as possible.

julek-wolfssl commented on src/tls.c:
```diff
@@ -8218,6 +8215,45 @@ static int TLSX_SupportedGroups_Find(WOLFSSL* ssl, word16 name)
+     * entries. The presence of this extension signals that the peer can be
+     * negotiated with. */
+    extension = TLSX_Find(*extensions, TLSX_KEY_SHARE);
+    if (extension == NULL) {
```
Here we are acknowledging that the peer sent a KeyShare extension. 0-len extension is also allowed as it means the peer want to ask what we support.

julek-wolfssl replied:
This is important in `RestartHandshakeHashWithCookie` as we add the extension when re-creating the HRR. We also error out in `DoTls13ClientHello` if it is not present.

rizlik replied:
I think this extension can't be `NULL`.
If invoked from `TLSX_PARSE->KS_PARSE`, we have the extension
If invoked from SendStatelessReply then we have a check that the extension exists (`tlsx.size != 0`). I don't think we should create the extension if it doesn't exist.

julek-wolfssl replied:
The existence of this extension is used deeper inside the stack. It should be added if we have the extension present (even if empty).

julek-wolfssl commented on src/tls.c:
```diff
@@ -8619,8 +8625,11 @@ int TLSX_KeyShare_Use(WOLFSSL* ssl, word16 group, word16 len, byte* data,
+        /* Generate a key pair. Casting to non-const since changes inside are
+         * minimal but would require an extensive redesign to refactor. Also
+         * this path shouldn't be taken when parsing a ClientHello in stateless
+         * mode. */
```
Because in stateless case we enter through `TLSX_KeyShareEntry_Parse` and we always provide a `data` parameter.

rizlik replied:
what if the client provided empty key share entries array?

julek-wolfssl replied:
The `data` parameter is always allocated in `TLSX_KeyShareEntry_Parse` except for the PQ case.

julek-wolfssl committed:
Define `usePSK` when ext is present

rizlik commented on src/dtls.c:
```diff
@@ -557,10 +557,10 @@ static int SendStatelessReply(const WOLFSSL* ssl, WolfSSL_CH* ch, byte isTls13,
+                else if ((modes & (1 << PSK_KE)) == 0) {
                         ERROR_OUT(PSK_KEY_ERROR, dtls13_cleanup);
                 }
+                usePSK = 1;
```
shouldn't we check which psk we'll use? now we are setting this to true even if the client provides an empty extension. Is it correct?

julek-wolfssl replied:
Refactored to only be set when we are sure the the client wants and can do PSK.

julek-wolfssl committed:
Fix Jenkins errors

julek-wolfssl committed:
Use PSK callback to get the ciphersuite to use

- Allocate additional byte in TLSX_PreSharedKey_New for null terminator

julek-wolfssl committed:
Rebase and Jenkins fixes

julek-wolfssl committed:
Re-create hs header for hash

julek-wolfssl committed:
Refactor SendStatelessReply 1.3 branch into new function

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+        else
+            ret = SendAlert((WOLFSSL*)ssl, alert_fatal, illegal_parameter);
+#else
+#error "WOLFSSL_SEND_HRR_COOKIE has to be defined to use DTLS 1.3 server"
```
Moved outside.

julek-wolfssl committed:
Calculate cookie in SendStatelessReplyDtls13()

Not touching ssl->hsHashes while in stateless mode

rizlik commented on src/dtls.c:
```diff
@@ -250,23 +291,30 @@ static int TlsTicketIsValid(WOLFSSL* ssl, WolfSSL_ConstVector exts,
     if (ret != WOLFSSL_TICKET_RET_OK && ret != WOLFSSL_TICKET_RET_CREATE)
         return 0;
+    /* Store info for later */
+#if defined(WOLFSSL_TLS13) && defined(HAVE_SESSION_TICKET)
```
we shouldn't parse the TLSX_SESSION_TICKET in DTLS 1.3 because the extension isn't allowed.

julek-wolfssl replied:
Good catch. I also changed it so that the resumption info is checked for DTLS 1.3 since it affects cipher negotiation.

julek-wolfssl committed:
Fix unused variable

julek-wolfssl committed:
Definition was defined out in some configs

julek-wolfssl committed:
For DTLS 1.3 use PSK for ticket 

Resumption info is also necessary when WOLFSSL_DTLS_NO_HVR_ON_RESUME is not defined.

julek-wolfssl committed:
Missing casts

rizlik commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+            XMEMCPY(suites.hashSigAlgo, tlsx.elements + OPAQUE16_LEN,
+                    len);
+        }
+
```
Is `SIGNATURE_ALGORITHMS` extension obligatory if not using PSK? If so, should we error out?

julek-wolfssl replied:
Adding check and erroring out before `MatchSuite_ex`.

rizlik commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+            haveKS = 1;
+#endif
+        }
+
```
If the client didn't provide any `KEY_SHARE` extension (meaning no extension at all, not a 0-sized extension) we should error out in case of KE, not just provide a suggesting group in the `HRR`. `TlsxFindByType` may need some refactoring to distinguish between a 0-sized extension and a not-present one.

julek-wolfssl replied:
Refactor to handle sent 0-len extensions. The error logic is outsourced to the functions that parse the extensions so that len checks are done in one place.

rizlik commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                    goto dtls13_cleanup;
+                if ((modes & (1 << PSK_DHE_KE)) &&
+                        !ssl->options.noPskDheKe) {
+                    if (!haveKS)
```
a zero sized keyshare extension shouldn't be fatal.

julek-wolfssl replied:
Refactored to handle empty extensions but KeyShare should still have the two length bytes.

rizlik commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+        else
+#endif
+        {
+            ret = MatchSuite_ex(ssl, &suites, &cs, parsedExts);
```
In case the client isn't using PSK based handshake. Do we check that he provides a good keyshare? I don't see `TLSX_KeyShare_Choose` invoked in the non-PSK code path

julek-wolfssl replied:
`TLSX_KeyShare_Choose` is invoked deep in `MatchSuite_ex`.

rizlik reviewed (changes requested):
An incredible amount of work. I left some comments as I think some PSK and cookie check is mishandled. I think the PR can improve a lot in the testing. If my comments are sound, I think we should, at least, add checks for that issues.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -133,45 +134,82 @@ static int ReadVector16(const byte* input, WolfSSL_ConstVector* v)
+        word16  len;
+        if (ch->cookie.size < OPAQUE16_LEN + 1)
+            return BUFFER_E;
+        ato16(ch->cookie.elements, &len);
```
`ch->cookie` gets overwritten with the extension value or 0 if not found. But I agree that it may be better to separate these fields.

julek-wolfssl committed:
Read cookie extension into separate field

julek-wolfssl commented on src/dtls.c:
```diff
@@ -133,45 +134,82 @@ static int ReadVector16(const byte* input, WolfSSL_ConstVector* v)
+        ato16(ch->cookie.elements, &len);
+        if (ch->cookie.size - OPAQUE16_LEN != len)
+            return BUFFER_E;
+        ret = TlsCheckCookie(ssl, ch->cookie.elements + OPAQUE16_LEN,
```
I don't think it is necessary. After our conversation, the RFC doesn't specify this and the hash + IP address is enough to include in the cookie.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+    (void)pskInfo;
+#ifdef WOLFSSL_DTLS13
+    if (isTls13) {
+#ifdef WOLFSSL_SEND_HRR_COOKIE
```
Moved.

julek-wolfssl committed:
Refactor alerts into one location

Remove previous stateless code. Now all DTLS 1.3 stateless handling is done in dtls.c

julek-wolfssl committed:
CopyExtensions -> CopySupportedGroup

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+    return 0;
+}
+
+static int CopyExtensions(TLSX* src, TLSX** dst, void* heap)
```
Renaming.

julek-wolfssl committed:
Reuse ReadVector16

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+            if (ret != 0)
+                goto dtls13_cleanup;
+        }
+
```
Adding checks.

julek-wolfssl committed:
Adding checks for SigAlgs, KeyShare, and Supported Groups

julek-wolfssl committed:
TLS 1.3: hold decrypted ticket to check which ciphersuite matches

DTLS 1.3: Move stateless ticket decoding to FindPskSuiteFromExt

julek-wolfssl commented on src/dtls.c:
```diff
@@ -232,41 +271,57 @@ static int TlsxFindByType(WolfSSL_ConstVector* ret, word16 extType,
         return 0;
     XMEMCPY(tempTicket, tlsxSessionTicket.elements, tlsxSessionTicket.size);
-    ret = DoDecryptTicket(ssl, tempTicket, (word32)tlsxSessionTicket.size, &it);
+    ret = DoDecryptTicket((WOLFSSL*)ssl, tempTicket,
```
Reverted back to only <= 1.2. Doing 1.3 parsing in `FindPskSuiteFromExt`.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -232,41 +271,57 @@ static int TlsxFindByType(WolfSSL_ConstVector* ret, word16 extType,
-    byte* isValid)
+#if defined(WOLFSSL_DTLS13) || defined(WOLFSSL_DTLS_NO_HVR_ON_RESUME)
+#ifdef HAVE_SESSION_TICKET
+static int TlsTicketIsValid(const WOLFSSL* ssl, WolfSSL_ConstVector exts,
```
DTLS 1.3 logic now in `FindPskSuiteFromExt`. `PskInfo` only used for DTLS 1.3 resumption cipher negotiation.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+        pskInfo->cipherSuite = sess->cipherSuite;
+        pskInfo->namedGroup = sess->namedGroup;
+
+        pskInfo->isValid = 1;
```
Excluded.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
         return 0;
 #endif /* HAVE_SESSION_TICKET */
-    ret = TlsSessionIdIsValid(ssl, ch->sessionId, isValid);
+    ret = TlsSessionIdIsValid(ssl, ch->sessionId, pskInfo);
```
Excluded.

julek-wolfssl committed:
Differentiate between empty and missing extension

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
-        ret = SendHelloVerifyRequest(ssl, cookie, DTLS_COOKIE_SZ);
+#endif
+#ifdef WOLFSSL_DTLS_NO_HVR_ON_RESUME
+    if (pskInfo.isValid) {
```
Separated logic. `WOLFSSL_DTLS_NO_HVR_ON_RESUME` only applies to 1.2 now.

julek-wolfssl committed:
Check for shared ctx in all configs

julek-wolfssl committed:
Rebase fixes

rizlik commented on src/internal.c:
```diff
@@ -34535,142 +34615,267 @@ static int DoSessionTicket(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
+        WOLFSSL_START(WC_FUNC_TICKET_DO);
+        WOLFSSL_ENTER("DoClientTicket_ex");
+
+        decryptRet = DoDecryptTicket(ssl, psk->identity, psk->identityLen,
```
`DoDecryptTicket` can return `WC_PENDING_E` if `WOLFSSL_ASYNC_CRYPT` is enabled. How this error should be propagated? at the moment it's translated in `PSK_DECRYPT_FAIL`  and it will cause a failure

rizlik commented on src/internal.c:
```diff
@@ -34535,142 +34615,267 @@ static int DoSessionTicket(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
+
+        ret = DoClientTicketCheckVersion(ssl, psk->it);
+        if (ret != 0) {
+            psk->decryptRet = PSK_DECRYPT_FAIL;
```
This has the side effect that `CleanupClientTickets` will not zero out the decrypted ticket

julek-wolfssl replied:
Fixing.

rizlik commented on src/dtls.c:
```diff
@@ -290,60 +374,530 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                        ret = WOLFSSL_TICKET_RET_REJECT;
+                        break;
+                }
+                if (ret == WOLFSSL_TICKET_RET_OK) {
```
should we check for  `WOLFSSL_TICKET_RET_CREATE`  as well?

julek-wolfssl replied:
```
 * @return  WOLFSSL_TICKET_RET_CREATE when successful and a new ticket is to
 *          be created for TLS 1.2 and below.
```
This is our documentation regarding `WOLFSSL_TICKET_RET_CREATE`. It is also not used in the stateful `DoPreSharedKeys`.

rizlik commented on src/dtls.c:
```diff
@@ -290,60 +374,530 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+        cs.cipherSuite0 = pskInfo.cipherSuite0;
+        cs.cipherSuite = pskInfo.cipherSuite;
+
+        if (haveSG && !haveKS) {
```
should this be `doKE &&  (!haveSG || !haveKS)`

julek-wolfssl replied:
No
https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.8
```
   -  If containing a "supported_groups" extension, it MUST also contain
      a "key_share" extension, and vice versa.  An empty
      KeyShare.client_shares vector is permitted.
```

rizlik commented on src/dtls.c:
```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+#error "WOLFSSL_SEND_HRR_COOKIE has to be defined to use DTLS 1.3 server"
+#endif
+
+static int SendStatelessReplyDtls13(const WOLFSSL* ssl, WolfSSL_CH* ch,
```
It's ok for me as it is

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +374,530 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+            }
+#endif
+#ifndef NO_PSK
+            {
```
Correct. If we don't recognize it as a ticket then we try to pass it to `FindPskSuite`.

julek-wolfssl committed:
Clear decrypted ticket that failed checks in DoClientTicket_ex

julek-wolfssl committed:
Add RFC link to help understand constraints

julek-wolfssl committed:
This path in TLSX_KeyShare_Choose should not be taken normally

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +374,530 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                    &searched);
+            if (ret != 0)
+                goto dtls13_cleanup;
+            if (cs.clientKSE == NULL && searched)
```
Changing so that it will error out by default.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +374,530 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                        "Supported Groups.");
+            ERROR_OUT(INCOMPLETE_DATA, dtls13_cleanup);
+        }
+        ret = MatchSuite_ex(ssl, &suites, &cs, parsedExts);
```
Adding comment.

julek-wolfssl committed:
Add comment about keyshare negotiation

julek-wolfssl committed:
Refactor sequence number reset for DTLS into one function

julek-wolfssl committed:
Reset DTLS sequence number

julek-wolfssl committed:
Async fixes

julek-wolfssl commented on src/dtls.c:
```diff
@@ -290,60 +374,530 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+            byte searched = 0;
+            ret = TLSX_KeyShare_Choose(ssl, parsedExts, &cs.clientKSE,
+                    &searched);
+            if (ret != 0)
```
Async fixed. Tested with `./configure --enable-all --enable-dtls13 --enable-asynccrypt`.

rizlik reviewed (approved):
This was a huge amount of work.  Thanks a lot @julek-wolfssl :top:

rizlik commented:
assigning to @SparkiDev for a final merge review

SparkiDev commented:
I have no idea about the strategy you used to implement this and what decisions you made to end up where we are.

What gets changed in the wolfSSL to create the stateless reply?
The reply is a ServerHello/HelloRetry so please rename SendStatelessReply and like.
You make a copy of the WOLFSSL fields that get change in a SendTls13ServerHello, can we do the same before this?

Would it be better to add the digest to the CipherSuiteInfo rather than call SetCipherSpecs_ex()?

SparkiDev commented on src/dtls.c:
```diff
@@ -290,60 +396,534 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+    return 0;
+}
+
+static int CopySupportedGroup(TLSX* src, TLSX** dst, void* heap)
```
Belongs in tls.c.
TLSX_SupportedGroup_Copy()

julek-wolfssl replied:
Moving.

SparkiDev commented on src/tls.c:
```diff
@@ -4546,7 +4543,8 @@ static int TLSX_PointFormat_Parse(WOLFSSL* ssl, const byte* input,
 
 #if defined(HAVE_ECC) || defined(HAVE_CURVE25519) || defined(HAVE_CURVE448)
-int TLSX_ValidateSupportedCurves(WOLFSSL* ssl, byte first, byte second) {
+int TLSX_ValidateSupportedCurves(const WOLFSSL* ssl, byte first, byte second,
```
Why pass in CipherSuite when ecdhCurveOID is the only thing being set?

julek-wolfssl replied:
Passing in only pointer to `ecdhCurveOID`.

SparkiDev commented on src/tls.c:
```diff
@@ -4866,21 +4864,21 @@ int TLSX_ValidateSupportedCurves(WOLFSSL* ssl, byte first, byte second) {
     {
         key = 1;
-        ssl->ecdhCurveOID = defOid;
+        cs->ecdhCurveOID = defOid;
```
ssl->ecdhCurveOID used to get updated.
I can't see that happening in VerifyServerSuite anymore.
Is this intentional?

SparkiDev commented on src/tls.c:
```diff
@@ -4866,21 +4864,21 @@ int TLSX_ValidateSupportedCurves(WOLFSSL* ssl, byte first, byte second) {
     {
         key = 1;
-        ssl->ecdhCurveOID = defOid;
+        cs->ecdhCurveOID = defOid;
```
ssl->ecdhCurveOID was being updated before.
Now we check ssl->ecdhCurveOID without updating.
Is this correct behaviour?

julek-wolfssl replied:
Changing all checks to `*ecdhCurveOID` and setting initial value to `*ecdhCurveOID = ssl->ecdhCurveOID;`. This should replicate the previous behaviour.

SparkiDev commented on src/dtls.c:
```diff
@@ -206,17 +294,18 @@ static int ParseClientHello(const byte* input, word32 helloSz, WolfSSL_CH* ch)
-#ifdef HAVE_SESSION_TICKET
+#if (defined(WOLFSSL_DTLS_NO_HVR_ON_RESUME) && defined(HAVE_SESSION_TICKET)) \
+    || defined(WOLFSSL_DTLS13)
 static int TlsxFindByType(WolfSSL_ConstVector* ret, word16 extType,
```
I see. The types aren't from TLSX.
Don't name it Tlsx then.

Why isn't 'exts' being passed a pointer?

julek-wolfssl replied:
Changing function name. `exts` isn't passed in because this only looks for the raw extension data and doesn't do any extension parsing.

julek-wolfssl commented on src/keys.c:
```diff
@@ -44,10 +44,81 @@
+}
+
+/**
+ *
```
Adding comment.

julek-wolfssl committed:
Address code review

julek-wolfssl commented:
>What gets changed in the wolfSSL to create the stateless reply?

We recreate a minimal version of `DoClientHello` and `DoTls13ClientHello`. `DoClientHello` is easy to recreate because we just do a sanity check and check if the cookie matches. If it doesn't then request a HelloVerifyRequest. `DoTls13ClientHello` is hard because we need to also do the cipher and curve negotiation in one HelloRetryRequest.

>The reply is a ServerHello/HelloRetry so please rename SendStatelessReply and like.

>You make a copy of the WOLFSSL fields that get change in a SendTls13ServerHello, can we do the same before this?

Do you mean for the entire `WOLFSSL` object? `DoClientHello` and `DoTls13ClientHello` do a lot of manipulation and allocation when parsing the ClientHello. @rizlik and I considered this but we change too much when parsing.

>Would it be better to add the digest to the CipherSuiteInfo rather than call SetCipherSpecs_ex()?

I assume you mean `CipherSuite`. I wanted to make minimal necesary changes to parsing logic. `CreateCookie` uses  `ssl->specs.mac_algorithm` and setting the `ssl->specs` makes it future proof when new changes are introduced that use `ssl->specs`.

julek-wolfssl commented:
Jenkins retest this please

SparkiDev commented:
I considered copying the wolfSSL object as an alternative design but we don't have a wolfSSL_copy() and it would require a lot more memory usage.

I don't necessarily like all the extra work we are doing but I spent some time looking through the changes and I can see why.

Thanks,
Sean

dgarske reviewed (approved):
Over to PR Captain @JacobBarthelmeh

JacobBarthelmeh commented:
Am running some fuzz testing on these changes.
